### PR TITLE
feat(mobile): reset a spray wall and keep its climbs findable (SW-13)

### LINKED ISSUE #5446: SW-13: Reset mobile: new photo, compare, confirm; badge, filter, remix
Epic #5346 (SW-13). The reset on the phone: new photo, anchors, compare kept / removed / new, confirm; then the "lost holds" badge, the filter, and Remix.

## Why

This is the half of the feature that makes a spray wall a living board instead of a snapshot. The review has to be honest (the matcher is heuristic) and cheap (one screen, every ring toggleable), and a broken climb must stay findable by default and playable, with a one-tap path to remix it onto the current holds.

## Scope

- "New photo" in `BoardDetailSheet` → `app/boards/spray/reset.tsx` (reuses the SW-09 photo step) → anchors are **required** here: if the previous version has none, the user taps them on the old photo first; the previous alive holds are overlaid through the composite homography as a tapping guide → upload → `createSprayWallVersion` → on-device detection (SW-09 pipeline) → `proposeSprayWallReset`. If the new photo's aspect ratio differs from the previous version's by more than 20 %, show a "Same wall?" warning with Continue / Retake; never block (owner decision 2026-09-14).
- `app/boards/spray/compare.tsx` = the SW-08 editor in compare mode: rings colour-coded kept / removed / new / low-confidence, a segmented filter, header "12 kept · 3 removed · 5 new · 4 climbs lose holds", every ring toggleable (a removed ring can be "still here", a new ring can be rejected, a low-confidence match can be split), "same hold, moved here" pairing from `movesSuggested`, manual add / move / resize on new holds; confirm → `commitSprayWallVersion` → refresh the active board, invalidate the registry, the render caches (version-keyed per SW-07) and the climb queries.
- Climb list: `NeedsRepairChip` following the `HiddenChip` pattern in `ClimbListItemContent.tsx`, from `missingHoldCount`; `holdIntegrity` filter in `ClimbFilterSheet.tsx` (Collection section) + `packages/shared/climb-filters/src/filter-state.ts`; **default = All with the chip** (findable by default; one tap to "Intact only"); broken climbs stay playable.
- Play drawer banner "3 holds are gone" with a Remix CTA; `LostHoldsLayer.tsx` draws dashed ghost rings at the removed holds' stored positions (`Climb.lostHolds`) through `InteractiveCreateBoard`'s `overlay` slot and in the play view.
- **Remix** = the existing `openRemix` (`use-create-climb-navigation.ts`) with `lostHoldIds`: the `initHistory` sanitizer already drops ids not on the board, so the editor opens with the surviving holds painted and the ghosts visible; successor suggestions from `movedFromHoldId` pre-painted as HAND with a hint; save records lineage via `remixClimb` (SW-12). The child shows its parent ("Remixed from <name>", tappable to the parent's climb view with the parent's ticks and grade history); the parent may no longer be climbable because of moved or removed holds, and that is fine — it keeps its own history and chip (owner decision 2026-09-14).
- Telemetry: `Spray Wall Reset Previewed`, `Spray Wall Reset Applied`, `Climb Remixed From Broken` (SW-17 names).

- **One active draft per wall** (epic decision 2026-09-15): the editor works on THE draft; "New photo" / "Edit holds" while a draft exists resumes that draft rather than creating another; a Discard action ends it. Publishing an older generation over a newer one is rejected by the server.

## Out of scope

Feature-matching alignment, polygon vertex editing, remix history UI beyond a "remixed from" line.

## Acceptance

- QA plan on a seeded wall: take a photo with 2 holds removed and 1 added → compare shows 2 removed / 1 new with sensible kept rings → confirm → the two affected climbs show the chip → the filter "Lost holds" finds them → open one → ghost rings at the missing holds → Remix → editor opens with the surviving holds → save → the new climb has no chip and shows "remixed from".
- Rejecting a removed ring in compare keeps the hold alive after commit; a low-confidence match can be split into removed + new.
- Everything ships by OTA.

## Where

`packages/mobile/app/boards/spray/reset.tsx`, `compare.tsx` (new), `packages/mobile/src/components/outline-editor/*` (compare mode), `packages/mobile/src/components/climb-list/ClimbListItemContent.tsx`, `ClimbFilterSheet.tsx`, `packages/shared/climb-filters/src/filter-state.ts`, `packages/mobile/src/components/create-climb/LostHoldsLayer.tsx` (new), `InteractiveCreateBoard.tsx`, `use-create-climb-navigation.ts`, `PlayDrawer`.

## Blocked by / Blocks

- Blocked by: SW-08 (#5441), SW-09 (#5442), SW-12 (#5445).
- Blocks: SW-17 (#5450).

## Size

L.

### PR DESCRIPTION
**Pull request by marcodejongh**
## Summary

The reset half of spray walls (epic #5346, SW-13): a wall owner photographs the wall again, marks its corners, reviews what changed ring by ring, and confirms. Climbs that lost holds stay findable, stay playable, get a badge, and can be remixed onto what is on the wall now.

Stacked on #5484 (SW-09) → #5481 (SW-08) → #5479 (SW-07), and merges #5480 (SW-12); merge those first, then this PR retargets to main.

**The reset flow** (`/boards/spray/reset?wallUuid=…`). One route with a pure stepper behind it (`reset-wall-machine.ts`), a sibling of SW-09's add-a-wall machine rather than a branch inside it — because the two disagree about the one thing that matters most in either: **anchors are mandatory here**. The canonical frame is version 1's photo frame forever, so from version 2 on the four corners are the only thing that says where the new photograph sits in it; without them the identity homography claims the second photo has the same crop as the first, the detections arrive labelled with coordinates from another picture, and the matcher reports the whole wall gone. `ANCHORS_DONE` does nothing without a usable convex quad, so a climber never uploads a photograph the server is going to refuse.

**The compare view** is the SW-08 editor's reasoning applied to a proposal rather than a second editor: the draft photo under `InteractiveFilterBoard`, rings colour-coded AND dash-coded (kept / gone / new / unsure — a reset is decided from these rings, so a colour-blind owner has to be able to decide it), a segmented filter, a counts header, every ring toggleable, and "same hold, moved here" pairing. A pairing is unrepresentable unless the server would take it: `canPairMove` is the only way a `movedFromHoldId` enters the review, and putting a predecessor back on the wall drops every pairing naming it. After `commitSprayWallVersion` lands, `invalidateSprayWallRenderData` re-registers the wall immediately — this device is the one that moved the version token every spray cache key is built on.

It is a component, not a second route, and the detections are why: a wall photo yields hundreds of circles, expo-router params are strings, and the proposal is expressed in INDICES into that one array. A route handoff would mean a module-level stash — a second source of truth for the array the whole review is numbered against.

**No detections, no review.** The matcher compares two sets of circles, so an empty second set honestly means "the whole wall has gone" — which, committed, breaks every climb on it. A phone with no detector lands in the hold editor on the add-a-wall flow and is fine; there is no equivalent fallback for a reset, so the screen says so and offers nothing else.

**A climb that lost holds** keeps its place: a "2 holds gone" chip on the climb row (neutral grey, beside the Hidden chip — colour in that row means grade and nothing else), a **Holds** filter in the filter sheet defaulting to All, and a play-drawer banner that states the number and offers Remix through the same `useCreateClimbNavigation` handoff the actions sheet uses. `missingHoldCount` now rides the search documents, the queue types and the party wire, so a peer standing at the same wall sees the same explanation for the same gaps.

**Remix opens on the holds that are actually there.** The seed is the parent's frames, which still name every hold the climb was set on — including the ones a reset took off — and nothing downstream was dropping them: `filterSupportedFrames` asks whether the BOARD TYPE supports a hold's STATE, never whether the hold is on the wall. So `useCreateClimb` gains an opt-in `availableHoldIds`, and the mobile create screen passes the wall's live hold ids from the SW-07 registry for `spray` and nothing at all for a catalogue board, where every hold in a seed is real by construction. Without it the editor opened with the lost holds painted — no placement so no ring, no target so no way to tap them off — while they still counted toward `startingCount`, `finishCount` and `isValid`, and Save published a climb born broken on the one flow that exists to repair one.

`Climb.lostHolds` (backend + SDL + resolver + tests) carries the geometry of those holds as they were, for ghost rings. The layer that draws them is not in this PR: the play drawer renders through the board-render pipeline rather than `InteractiveCreateBoard`'s `overlay` slot, and the create editor would need the parent's holds threaded through string route params — both are their own piece of work, and the banner already answers the question the missing holds raise. Follow-up: #5493.

Author-side verification: `vp run typecheck`, `vp run typecheck:mobile`, `vp check` (0 errors), `vp test run --project mobile` (full suite), `vp test run --project i18n`, `vp run check:i18n`, `vp run check:mobile-bundle`, and the backend `climb-lost-holds` / `spray-wall-reset` files individually. The anchors gate and the integrity chip were each mutation-tested (flip the predicate, watch a test fail, restore).

## Release Notes

- Reset your spray wall from your phone: take a new photo, mark the corners, and check what the app thinks came off before anything changes.
- Every ring is yours to argue with — put a hold back, reject a false one, or say "same hold, moved here".
- Climbs that lost holds stay in your list with a badge instead of vanishing, and there's a new Holds filter when you only want the intact ones.
- Opened a climb that lost holds? One tap to remix it, and the editor opens with only the holds still on the wall — the missing ones are gone from the copy, not carried into it.

## Test plan

Reachable from the wall's page once #5491 lands; until then the route is exercised by the tests.

1. Boards → your spray wall → New photo.
2. Pick a photo. Next is blocked until four corners are set.
3. Drag the corners, tap Use these corners, wait.
4. Compare screen: tap a red ring → "It's still here".
5. Confirm the reset → Climbs list shows a "holds gone" badge.

## Risk

Risk: 4/5 — reset flow writes a new wall generation

Closes #5446

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Q7BjHrRALJ1gjtiRWm5rLV

### CODE REVIEW

File: packages/backend/src/graphql/resolvers/climbs/lost-holds.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+ * climber on the wall ghost rings for holds that are still bolted on, the moment
+ * the owner started a reset and walked away — and nothing would ever take them
+ * back off.
+ *
+ * `spray_wall_holds.hold_id` is globally unique across walls (both it and the
+ * catalogue ids it doubles as come from the single `spray_hold_catalog_id_seq`
+ * — see `allocateHoldIds`), so joining on `hold_id` alone cannot pick up another
+ * wall's hold and no layout scope is needed on top.
+ */
+export async function resolveClimbLostHolds(climb: ClimbLostHoldsParent): Promise<GraphQLLostHold[] | null> {
```
[the remark this answers is not in the document]
Comment (marcodejongh) [reply packages/backend/src/graphql/resolvers/climbs/lost-holds.ts]: Right, and worse than the report says: the resolver never received a context at all, so there was nothing to check against even if it had wanted to.

Fixed in `7b4b95cbe`. Nothing on the parent is trusted now — the climb's own `board_climbs` row is joined for its `layout_id`, and `sprayClimbVisibilityCondition` (the same by-layout rule every other spray climb reader carries: owner, gym member, or a public wall) goes in the WHERE beside it, with the viewer id from `ctx`. One statement rather than two, so there is no window between deciding and reading, and a caller who may not see the wall gets an **empty list** — byte for byte what an intact climb on a visible wall returns, so the field is not an oracle for which climb uuids belong to private walls. The parent's `missingHoldCount` survives only as a fast path; each of those can withhold rows and none can produce any.

Written as a SQL predicate rather than a call to `viewerCanSeeSprayWallByLayout` deliberately: SW-17 changes that helper's signature on its own branch, and this resolver is what both branches have to merge.

Tests in `packages/backend/src/__tests__/climb-lost-holds.test.ts`, on the same private-wall fixture, using exactly the forged parent you describe (`boardType: 'spray'`, `missingHoldCount: 99`): a signed-in stranger gets `[]`, an anonymous caller gets `[]`, a null context gets `[]`, the owner still gets both holds with their geometry, and once the owner makes the wall public the stranger sees the same rows. Mutation: deleting the `sprayClimbVisibilityCondition` clause turns the stranger's case red (`expected [ …(2) ] to deeply equal []`); restored and re-run green. Run against a private Postgres, 4 passed.

File: packages/mobile/src/components/spray-wall/SprayWallResetScreen.tsx [hunk 2]
```diff
@@ -0,0 +1,667 @@
...
+   * so an ungated listener here would keep eating back presses on every screen
+   * pushed on top of this one.
+   */
+  const isFocused = useIsFocused();
+  useEffect(() => {
+    if (!isFocused) return;
+    const subscription = BackHandler.addEventListener('hardwareBackPress', () => {
+      goBack();
+      return true;
+    });
```
[the remark this answers is not in the document]
Comment (marcodejongh) [reply packages/mobile/src/components/spray-wall/SprayWallResetScreen.tsx]: Agreed, and the consequence is the sharp part: the draft is on the server but the detections that make it reviewable are not — they were computed from a local photograph — so a silent exit strands a draft nothing can resume and the only way out is discarding it and shooting the wall again.

Fixed in `7b4b95cbe`, mirroring `SprayWallWizardScreen`: a `beforeRemove` listener on the same `shouldConfirmLeave` predicate the footer uses (`isBusy || leavingKeepsDraft`), re-dispatching `event.data.action` on confirm so the exit the climber chose is the exit they get. It **replaces** the `BackHandler` listener rather than sitting beside it — two guards on one gesture would ask twice — so header, gesture and hardware back now share one behaviour.

The predicate is tested in `reset-wall-machine.test.ts`: no question before the draft exists, a question once `DRAFT_CREATED` has landed, a question mid-upload and mid-detection even though no draft exists yet (`leavingKeepsDraft` is false there and `shouldConfirmLeave` is still true), and no question after `COMMITTED`. 18 passed.

File: packages/mobile/app/boards/_layout.tsx [hunk 3]
```diff
@@ -44,6 +44,11 @@ export default function BoardsLayout() {
           rule 3 keeps off a sheet's own drag. Flag-gated inside the screen — the
           route existing is not the same as the feature being reachable. */}
       <Stack.Screen name="spray/new" options={{ title: tBoards('sprayWizard.screenTitle') }} />
+      {/* Resetting a wall — a new photograph of a wall that already carries
+          climbs. Same route-not-sheet reasoning as the flow above: the corner
+          markers and the compare view are both full-screen pan-and-pinch
+          surfaces. SW-11 adds the "New photo" row that navigates here. */}
+      <Stack.Screen name="spray/reset" options={{ title: tBoards('sprayReset.screenTitle') }} />
```
[the remark this answers is not in the document]
Comment (marcodejongh) [reply packages/mobile/app/boards/_layout.tsx:60]: The search result is correct and the conclusion is the one thing I want to push back on: the entry point is not missing, it is deliberately not mine to add.

The "New photo" row belongs to the wall's own page — SW-11's `BoardDetailSheet` rows, still dark behind `SPRAY_DETAIL_ROWS_ENABLED = false` — which lives on a **different stack** from `/boards`. #5491 (SW-11b) wires that row to this route once both stacks have merged. Adding an owner-facing action here would be a duplicate the day #5491 lands, and the worse of the two: the wall uuid this route needs is on the detail sheet, not on the boards stack, so an entry point here would have to re-resolve it.

So the code change in `7b4b95cbe` is to the comment, which was the actual defect — it claimed a row already navigates here. It now states that nothing pushes the route yet, why (the other stack, the flag, #5491), and that until then the route is exercised by its tests and reachable by deep link. The PR body's Test plan gained the same caveat above its five steps: "Reachable from the wall's page once #5491 lands; until then the route is exercised by the tests."

**Comment by marcodejongh** [comment]
## Review items addressed

Head `f49db0520`.

### Tests added (`f49db0520`)

**`use-spray-wall-reset.ts`** — `packages/mobile/src/lib/spray/__tests__/use-spray-wall-reset.test.tsx`, 6 cases. Propose success → the proposal reaches state with the document and input asserted. Propose network failure → `isError`, `data` undefined, no retry, and nothing offered to commit: an empty `kept`/`added` is the matcher saying the whole wall has gone, so a failed proposal must not read as a proposal. Commit success → `invalidateSprayWallRenderData(client, wallUuid, layoutId)` called once plus the `searchClimbs` invalidation. Commit failure → rejects, invalidates nothing, and never retries on its own, because a reset that appears to have failed may have landed.

**Both screens** — `SprayResetCompareScreen.test.tsx` (9) covers the four view states, a ring tap reaching the reducer, bulk-accepting the suggested moves, Confirm sending the reviewed payload, the aspect-mismatch warning not blocking, and the error banner on a failed propose. `SprayWallResetScreen.test.tsx` (8) covers loading, not-yours, nothing-published, the unresumable-draft discard, the photo step, a cancelled picker, and the anchors gate in both directions (four corners enable it; a bow tie does not).

**`SprayResetSvgLayer.tsx`** — 7 cases: a distinct colour **and** a distinct dash per kept/removed/added/lowConfidence (a reset decided from colour alone is one a colour-blind owner cannot make), a hold following its verdict across a toggle, the filter reaching the drawing rather than only the tap targets, a matched detection never drawn on top of the hold it is, the selection ring on both sides of the `-(index + 1)` key, and nothing drawn before the board has a size.

### Not mine: the `remixClimb` successors SQL

Adding `AND removed_version_id IS NULL` to the successors query in `packages/backend/src/graphql/resolvers/board/spray-walls.ts` is SW-12's code (#5480), not this branch's — it arrives here through the merge. The JS `.filter((holdId) => alive.has(holdId))` is correct today, so this is a load-fewer-rows improvement on a wall with many resets. Leaving it to SW-12's implementer rather than editing their file across a stack boundary.

Same for the earlier round's **server-side uniqueness on `movedFromHoldId` across additions**: SW-12 added exactly that refine in `4088648f1`, and `6e9775c55` here makes the client payload satisfy both new `added` refines by construction (`canPairMove` for one-predecessor-per-addition, and collapsing duplicates that round to the same canonical centre and radius — rounding to integers can merge two blobs a pixel apart in a 2048 px photo).

### Intentional, not gaps

- **`/boards/spray/reset` has no in-app entry point.** Staged by design: the "New photo" row lives on the wall's own page (SW-11's `BoardDetailSheet`, dark behind `SPRAY_DETAIL_ROWS_ENABLED = false`) on a different stack, and **#5491** wires it once both stacks merge. A second entry point here would be a duplicate the day that lands, and a worse one — the wall uuid is on the detail sheet, not on this stack. The route comment in `_layout.tsx` now says so, and the Test plan carries the caveat.
- **`eslint-disable react-hooks/exhaustive-deps` in `use-create-climb-screen.ts`** is the deliberate seed-once pattern, matching `initialFrames` directly above it, with the reason in the docblock: re-reading the wall's holds later would erase a hold the climber had just painted if a reset landed on another device.
- **`screenshot-locale.test.ts`'s 30 s timeout** buys module load for the whole catalogue graph under `resetModules`, not a hidden regression.
- **Offline + `holdIntegrity`** returning an empty list under an active chip is the known SW-15 (#5448) limitation — the column is not on the device yet, and declining is the faithful mirror. A user-visible fallback message is worth a follow-up; flagging rather than adding it here.

### Gates

`typecheck:mobile` ✅ · reset suites **67** · spray suites **31 files / 444** · new files **6 + 9 + 8 + 7 = 30** · `vp check` **0 errors** · `check:mobile-bundle` ✅ iOS + Android.

**Comment by marcodejongh** [comment]
## Review round addressed — head `828a94365`

### Fixed

**`LostHoldsBanner.tsx`** — the no-op `useCallback(() => onRemix?.())` wrapper is gone; the Button renders only when `onRemix` exists and passes it no arguments, so `onPress={onRemix}` is the same call with one closure fewer. Good catch.

**Three compounding per-tap costs in the compare view**, found applying the mobile performance checklist to a capped wall (1,500 holds, up to ~3,000 rings, re-rendered on every verdict toggle):

- `holdRingRole` ended in `lowConfidenceHoldIds.includes(holdId)` — an O(n) scan called once per hold inside two per-hold loops (the SVG layer's bucketing and `buildResetRingTargets`). Now an O(1) Set cached on the array's identity in a `WeakMap`: the role is a pure function of state rather than a hook, so no caller can memoise it, and the array is built once per proposal and never replaced by a transition. `SprayHoldSvgLayer` already reaches for a Set at this scale.
- The bucketing memo listed `review`, so every `[...outline]` copy, `radiusRingToBoardPx` and `ringToPathData` re-ran for all ~3,000 rings on each toggle, pairing and filter change — but ring geometry depends only on where a hold **is**, and a verdict never moves one. Path strings now build once per wall and a toggle only concatenates strings that already exist. The selection ring reads the same two caches instead of rebuilding its own.
- The five `strokeDasharray={[dash, …]}` literals were a fresh array identity for react-native-svg to diff on every tap, because `selectedKey` moves. Memoised on the board width, matching the sibling layer's documented reasoning.

No behaviour change — 10,608 mobile tests pass unchanged.

### Confirmed rather than changed

**Climbs stay findable after a reset** — traced end to end, not taken on the description's word. A removal only *stamps* `spray_wall_holds.removed_version_id`; nothing is deleted, the superseded version row stays, and `board_climbs` / `board_climb_holds` / the removed holds' `board_holes` + `board_placements` rows are all untouched, so a climb's frames still resolve to real catalogue rows. On the read side `holdIntegrity` defaults to `undefined` → `ANY` → **no predicate at all** (`holdIntegrityCondition`), so a broken climb is returned by an unfiltered search exactly as before; `CACHE_VERSION` moving v8 → v9 is what stops a cached page serving it without the badge.

**`is_listed = false` survives a reset.** The commit writes no row that has the column — `board_holes`, `board_placements` and `spray_wall_holds` have none — and the one `is_listed`-bearing table it touches, `board_product_sizes_layouts_sets`, is updated with `imageFilename` alone, so the seeded `false` is never in the `.set()`.

**The hold cap is server-side.** `nextTotal = existing.length - removed.length + added.length` against `MAX_HOLDS_PER_WALL`, computed from `aliveHolds` read under the wall lock, before `allocateHoldIds` and before any insert — and the subtraction can't be inflated, since `removed` is deduped by the schema refine and every id in it is proven alive first.

**Nothing leaks into the five excluded surfaces.** No `packages/web` file changes, none of the five gatekeepers is in the diff, and `isCacheableBoard` still excludes `spray`. The new readers are gated: `Climb.lostHolds` puts `sprayClimbVisibilityCondition` in the same WHERE as the geometry and resolves `layout_id` from the climb's own row, so a forged `ClimbInput` parent buys nothing and a stranger gets the same empty list an intact climb returns.

### Flagged, not fixed

**`resetReviewCounts` can overstate "N new"** by the number of accepted detections that round to the same canonical centre and radius, which `buildResetCommitDecisions` collapses. Sub-pixel edge case (two blobs a pixel apart in a 2048 px photo), the commit result reports the server's counts, and fixing it means threading `detections` through a signature used by the screen and its tests — noting it rather than widening this PR.

**`lostHolds` N+1** is real but not reachable: it is in `CLIMB_DETAIL_FIELDS` only, and `missingHoldCount` short-circuits the common cases without a query. A dataloader belongs with the ghost-ring work in #5493.
